File: src/BusinessLogic/Handlers/InlineQueryHandlers/RelatedInlineQueryHandler.cs
using Bounan.Bot.BusinessLogic.Clients.Shikimori;
using Bounan.Bot.BusinessLogic.Clients.Shikimori.Models;
using Bounan.Bot.BusinessLogic.CommandDto;
using Bounan.Bot.BusinessLogic.Configs;
using Bounan.Bot.BusinessLogic.Constants;
using Bounan.Bot.BusinessLogic.Extensions;
using Bounan.Bot.BusinessLogic.Helpers;
using Bounan.Bot.TelegramBot.Telegram;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using Telegram.Bot.Requests;
using Telegram.Bot.Types;
using Telegram.Bot.Types.InlineQueryResults;

namespace Bounan.Bot.BusinessLogic.Handlers.InlineQueryHandlers;

public class RelatedInlineQueryHandler(
    ILogger<RelatedInlineQueryHandler> logger,
    IShikimoriApi shikimoriApi,
    IOptions<ShikimoriConfig> shikimoriConfig) : IInlineQueryHandler
{
    public static bool CanHandle(InlineQuery item) => item.Query.StartsWith(RelatedCommandDto.Command);

    public async Task<AnswerInlineQueryRequest> HandleAsync(
        InlineQuery inlineQuery,
        CancellationToken cancellationToken)
    {
        logger.LogInformation("Handling related inline query {Query}", inlineQuery.Query);

        var commandDto = CommandConvert.DeserializeCommand<RelatedCommandDto>(inlineQuery.Query);
        if (commandDto is null)
        {
            logger.LogWarning("Failed to deserialize command {Command}", inlineQuery.Query);
            return new AnswerInlineQueryRequest(inlineQuery.Id, []);
        }

        RelatedItem[] related;
        try
        {
            related = await shikimoriApi.GetRelatedAnimeAsync(commandDto.MyAnimeListId, cancellationToken);
            logger.LogInformation("Got related anime for {MyAnimeListId}", commandDto.MyAnimeListId);
        }
        catch (Exception)
        {
            related = Array.Empty<RelatedItem>();
            logger.LogWarning("Failed to get related anime for {MyAnimeListId}", commandDto.MyAnimeListId);
        }

        var relatedAnime = related
            .Where(rel => rel.Anime is not null)
            .ToArray();

        if (relatedAnime is null or { Length: 0 })
        {
            logger.LogInformation("No related anime for {MyAnimeListId}", commandDto.MyAnimeListId);
            return new AnswerInlineQueryRequest(
                inlineQuery.Id,
                new[]
                {
                    new InlineQueryResultArticle(
                        KnownInlineAnswers.NoRelatedAnime,
                        KnownInlineAnswers.NoRelatedAnime,
                        new InputTextMessageContent(KnownInlineAnswers.NoRelatedAnime)),
                });
        }

        var results = relatedAnime
            .OrderByDescending(rel => rel.Anime!.Id)
            .Select(anime => TelegramInlineItemsHelpers.AnimeInfo(
                anime.Anime!,
                shikimoriConfig.Value.BaseUrl,
                anime.RelationRussian.Or(anime.Relation)));

        logger.LogInformation(
            "Returning {Count} related anime for {MyAnimeListId}",
            relatedAnime.Length,
            commandDto.MyAnimeListId);
        return new AnswerInlineQueryRequest(inlineQuery.Id, results);
    }
}

File: src/handlers/on-webhook/event-handlers/inline-query-handlers/related-inline-query-handler.ts
import { Related } from '@lightweight-clients/shikimori-graphql-api-lightweight-client';
import { InlineQuery, InlineQueryResultArticle } from '@lightweight-clients/telegram-bot-api-lightweight-client';

import { getRelated } from '../../../../api-clients/cached-shikimori-client';
import { InfoCommandDto, RelatedCommandDto } from '../../command-dtos';
import { KnownInlineAnswers } from '../../constants/known-inline-answers';
import { InlineQueryHandler } from '../query-handler';

const canHandle = (inlineQuery: InlineQuery): boolean =>
    inlineQuery.query?.startsWith(RelatedCommandDto.Command) ?? false;

const handler: InlineQueryHandler = async (inlineQuery) => {
    console.log('Handling dubs inline query {Query}', inlineQuery.query);

    const commandDto = RelatedCommandDto.fromPayload(inlineQuery.query) as RelatedCommandDto;
    if (!commandDto) {
        console.warn('Failed to deserialize command', inlineQuery.query);
        return [];
    }

    const relatedRes = await getRelated(commandDto.myAnimeListId);
    const relatedAnimes = relatedRes.flatMap(x => x.related).filter(x => x?.anime) as Related[];
    const items: InlineQueryResultArticle[] = relatedAnimes.length === 0
        ? [
            {
                type: 'article',
                id: KnownInlineAnswers.NoRelatedAnime,
                title: KnownInlineAnswers.NoRelatedAnime,
                input_message_content: {
                    message_text: KnownInlineAnswers.NoRelatedAnime,
                },
            },
        ]
        : relatedAnimes.map(item => ({
            type: 'article',
            id: item!.anime!.id.toString(),
            title: item.anime!.russian || item.anime!.name,
            description: [item.relationText, item!.anime!.airedOn?.year].filter(x => !!x).join(', '),
            thumbnail_url: item.anime?.poster?.originalUrl,
            input_message_content: {
                message_text: new InfoCommandDto(item.anime!.id).toString(),
            },
        }));

    console.log('Returning {Count} related animes for {MyAnimeListId}', items.length, commandDto.myAnimeListId);
    return items;
}

export const relatedInlineQueryHandler = {
    canHandle,
    handler,
};
